nozzle/api_client.go

```diff
@@ -0,0 +1,65 @@
+package nozzle
+
+import (
+	cfclient "github.com/cloudfoundry-community/go-cfclient"
+)
+
+// APIClient wrapper for Cloud Foundry Client
+type APIClient struct {
+	clientConfig *cfclient.Config
+	client       *cfclient.Client
+}
+
+// AppInfo holds Cloud Foundry applications information
+type AppInfo struct {
+	Name  string
+	Space string
+	Org   string
+}
+
+// NewAPIClient crate a new ApiClient
+func NewAPIClient(apiURL string, username string, password string, sslSkipVerify bool) (*APIClient, error) {
+	config := &cfclient.Config{
+		ApiAddress:        apiURL,
+		Username:          username,
+		Password:          password,
+		SkipSslValidation: sslSkipVerify,
+	}
+
+	client, err := cfclient.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &APIClient{
+		clientConfig: config,
+		client:       client,
+	}, nil
+}
+
+// FetchTrafficControllerURL return Doppler Endpoint URL
+func (api *APIClient) FetchTrafficControllerURL() string {
+	return api.client.Endpoint.DopplerEndpoint
+}
+
+// FetchAuthToken wrapper for client.GetToken()
+func (api *APIClient) FetchAuthToken() (string, error) {
+	token, err := api.client.GetToken()
+	if err != nil {
+		return "", err
+	}
+	return token, nil
+}
+
+// ListApps wrapper for client.ListApps()
+func (api *APIClient) ListApps() map[string]*AppInfo {
+	appsInfo := make(map[string]*AppInfo)
+	apps, _ := api.client.ListApps()
```

Comment:
It's expensive to list all the apps each time we find a new/missing appId. There is an API to fetch the info for a single app: https://godoc.org/github.com/cloudfoundry-community/go-cfclient#Client.AppByGuid.

We should also use a cache that expires items after a certain time. go-cache seems a good fit: https://github.com/patrickmn/go-cache. But it doesn't have an upper bound on the number of items. We can handle that manually before we add to the cache. In our Java nozzle we cache up to 5000 apps and expire entries after 6 hours. We can do similarly here as well.

Reply:
is already cache where is used:
https://github.com/wavefrontHQ/cloud-foundry-nozzle-go/blob/b76bcf5f84deaa44cceebca68f58975853c27b1b/nozzle/nozzle.go#L85

But is not well done and I detected a bug yesterday... I will reimplement it.